### PR TITLE
[#17] Implement simple branching

### PR DESCRIPTION
**Pull request by ixahedron**
Problem: branching is done through structured instructions like `block` and branches targeting them, of which we have none.

Solution: implemented support for the `block` and `loop` instructions. This isn't _branching_ yet, strictly speaking, but it's necessary for it. Also added support for `if`, which _is_ branching, if in the most primitive of senses.

Follows #17

### CODE REVIEW

File: Main.lean [hunk 1]
```diff
@@ -376,8 +388,8 @@ def main : IO Unit := do
       (param i32)
       (result i32 i32) (result i32 i32)
 
-      (i32.const 1)
-      (nop)
+      (block (result i32) (i32.const 0))
+      (if (result i32) then (i32.const 8) else (i32.const 2))
```
Comment (cognivore) [inline Main.lean:392]: What does result mean here?

File: Wasm/Bytes.lean [hunk 2]
```diff
@@ -76,22 +83,30 @@ mutual
...
-    | .const (.i _) (.i ci) => ByteArray.mk #[0x41] ++ sLeb128 ci.val
+    | .const (.i 32) (.i ci) => b 0x41 ++ sLeb128 ci.val
+    | .const (.i 64) (.i ci) => b 0x42 ++ sLeb128 ci.val
     | .const _ _ => sorry -- TODO: float binary encoding
     | .add t g1 g2 =>
       -- Enter stackman
       extractGet' g1 ++ extractGet' g2 ++ extractAdd t
+    | .block ts ops =>
+      let bts := flatten $ ts.map (b ∘ ttoi)
+      let obs := bts ++ b ops.length.toUInt8 ++ flatten (ops.map extractOp)
```
Comment (cognivore) [inline Wasm/Bytes.lean:98]: Im 99% sure we need LEB everywhere. There can trivial ly be more than 255 instructions in else / then

File: Wasm/Bytes.lean [hunk 3]
```diff
@@ -76,22 +83,30 @@ mutual
...
+      b 0x02 ++ bts ++ lindex obs ++ b 0x0b
+    | .if ts thens elses =>
+      let bts := flatten $ ts.map (b ∘ ttoi)
+      let bth := b thens.length.toUInt8 ++ flatten (thens.map extractOp)
+      let belse := if elses.isEmpty then b0
+        else
+          let bel := b elses.length.toUInt8 ++ flatten (elses.map extractOp)
+          b 0x05 ++ lindex bel
+      b 0x04 ++ bts ++ lindex (bth ++ belse) ++ b 0x0b
```
Comment (cognivore) [inline Wasm/Bytes.lean:107]: Lgtm, but binary testing is very easy, we should be doing it even from main.

File: Wasm/Engine.lean [hunk 4]
```diff
@@ -143,45 +145,58 @@ mutual
...
-      | .num n0, .num n1 => match n0, n1 with
         | .i ⟨b0, i0⟩, .i ⟨_b1, i1⟩ => pure $ .num $ .i ⟨b0, i0 + i1⟩ -- TODO: check bitsize and overflow!
         | .f ⟨b0, f0⟩, .f ⟨_b1, f1⟩ => pure $ .num $ .f ⟨b0, f0 + f1⟩ -- TODO: check bitsize and overflow!
         | _, _ => throw .param_type_incompatible
       pure (res :: stack1)
+    | .block ts ops => do
+      -- TODO: currently, we only support simple [] → [valuetype*] blocks,
+      -- not type indices. For this reason, we start the block execution
+      -- with an empty stack to simulate 0-arity.
+      let go σ op := do runOp locals (←σ) op
```
Comment (cognivore) [inline Wasm/Engine.lean:168]: Lgtm except I can't read this \leftarrow notation again

File: Wasm/Engine.lean [hunk 5]
```diff
@@ -143,45 +145,58 @@ mutual
...
-  let locals := (f.params ++ f.locals).map $
-    fun l => match l.name with
-      | .some name => (.some name, pσ.2.get? l.index)
-      | .none => (.none, pσ.2.get? l.index)
-  let go (oσ : Except EngineErrors Stack) (x : Operation)
-         : Except EngineErrors Stack := do
-    let stack ← oσ
-    Stack.mk <$> runOp locals (Stack.es stack) x
+  let locals := (f.params ++ f.locals).map
+    fun l => (l.name, pσ.2.get? l.index)
```
Comment (cognivore) [inline Wasm/Engine.lean:198]: Always happy when the amou3if code reduces

File: Wasm/Wast/AST.lean [hunk 6]
```diff
@@ -81,12 +90,15 @@ mutual
...
 
+-- TODO: add support for function type indexes for blocktypes
 -- TODO: replace "NumUniT" with something supporting ConstVec when implemented
 -- TODO: generalise Consts the same way Get is generalised so that i32.const can't be populated with ConstFloat!
   inductive Operation where
   | nop
   | const : Type' → NumUniT → Operation
   | add : Type' → Get' → Get' → Operation
+  | block : List Type' → List Operation → Operation
+  | if : List Type' → List Operation → List Operation → Operation
```
Comment (cognivore) [inline Wasm/Wast/AST.lean:101]: Groovy

File: lakefile.lean [hunk 7]
```diff
@@ -1,7 +1,9 @@
 import Lake
 open Lake DSL
 
-package Wasm
+package Wasm {
+  precompileModules := true
```
Comment (cognivore) [inline lakefile.lean:5]: Uh oh

**Review by cognivore — CHANGES_REQUESTED** [review]
Check for LEB pls.

File: Main.lean (hunk 1 above)
Comment (ixahedron) [reply Main.lean:392]: The result type of the branches. It should actually support type indices, but for now it's just inlined value types. See [spec](https://webassembly.github.io/spec/core/syntax/instructions.html#syntax-blocktype)

Comment (cognivore) [reply Main.lean:392]: TIL. Great job!

File: Wasm/Engine.lean (hunk 4 above)
Comment (ixahedron) [reply Wasm/Engine.lean:168]: The left arrow notation is, AFAIU, a shorthand for `let sgm ← σ; runOp locals sgm op`

**Comment by ixahedron** [comment]
@cognivore added hopefully all necessary LEBbing, please rereview

**Review by cognivore — APPROVED** [review]

**Pull request merged**